subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "net_tls.h"
+#include "net_private.h"
+
+static int tls_tx(void *ctx, const unsigned char *buf, size_t len )
+{
+	struct net_context *context = ctx;
+	struct net_pkt *pkt;
+	int bytes, ret;
+
+	pkt = net_pkt_get_tx(context, K_FOREVER);
+
+	if (!pkt) {
+		return -EIO;
+	}
+
+	bytes = net_pkt_append(pkt, len, (u8_t *)buf, K_FOREVER);
+
+	ret = net_context_output(context, pkt, &context->remote);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+		return ret;
+	}
+
+	return bytes;
+}
+
+static int tls_rx(void *ctx, unsigned char *buf, size_t len )
+{
+	struct net_context *context = ctx;
+
+	ARG_UNUSED(context);
+
+	return -EOPNOTSUPP;
+}
+
+int net_tls_enable(struct net_context *context, bool enabled)
+{
+	int state;
+
+	if (!context) {
+		return -EINVAL;
+	}
+
+	if (context->options.tls == enabled) {
+		return 0;
+	}
+
+	state = net_context_get_state(context);
+	if (state != NET_CONTEXT_IDLE && state != NET_CONTEXT_UNCONNECTED) {
+		return -EBUSY;
+	}
+
+	if (enabled) {
+		mbedtls_ssl_init(&context->mbedtls.ssl);
+		mbedtls_ssl_config_init(&context->mbedtls.config);
+		mbedtls_ssl_set_bio(&context->mbedtls.ssl, context,
+				    tls_tx, tls_rx, NULL);
+	} else {
+		mbedtls_ssl_set_bio(&context->mbedtls.ssl,
+				    NULL, NULL, NULL, NULL);
+		mbedtls_ssl_config_free(&context->mbedtls.config);
+		mbedtls_ssl_free(&context->mbedtls.ssl);
+	}
+
+	context->options.tls = enabled;
+
+	return 0;
+}
+
+int net_tls_connect(struct net_context *context, bool listening)
+{
+	enum net_context_state state;
+	int err, role, type;
+
+	if (!context) {
+		return -EINVAL;
+	}
+
+	if (!context->options.tls) {
+		return 0;
+	}
+
+	state = net_context_get_state(context);
+	if (state != NET_CONTEXT_IDLE && state != NET_CONTEXT_UNCONNECTED) {
+		return -EBUSY;
+	}
+
+	role = listening? MBEDTLS_SSL_IS_SERVER: MBEDTLS_SSL_IS_CLIENT;
+
+	type = (net_context_get_type(context) == SOCK_STREAM)?
+		MBEDTLS_SSL_TRANSPORT_STREAM:
+		MBEDTLS_SSL_TRANSPORT_DATAGRAM;
+
+	err = mbedtls_ssl_config_defaults(&context->mbedtls.config, role, type,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (err) {
+		err = -ENOMEM;
+		goto mbed_err_conf;
```

Comment:
`mbedtls_ssl_session_reset` clears in/out buffers allocated in `mbedtls_ssl_setup`, therefore calling it from here will result in a fault. Actually, `mbedtls_ssl_setup` can also return an error without allocating the buffers, so error management should be refined a little bit.